Python/hashtable.c

```diff
@@ -109,10 +110,9 @@ round_size(size_t s)
     size_t i;
     if (s < HASHTABLE_MIN_SIZE)
         return HASHTABLE_MIN_SIZE;
-    i = 1;
-    while (i < s)
-        i <<= 1;
-    return i;
+    // 1 << _Py_bit_length(s) is the smallest value k so that 1 << k > s
+    // subtract one in case s is an exact power of two, saving space
+    return 1 << _Py_bit_length(s - 1);
```

Comment:
Can you please try _Py_SIZE_ROUND_DOWN() which could be even faster? It uses simple operations:

```
#define _Py_SIZE_ROUND_DOWN(n, a) ((size_t)(n) & ~(size_t)((a) - 1))
```

About _Py_bit_length(): it has a special case for 0, but s >= HASHTABLE_MIN_SIZE.

Reply:
That's brilliant! I didn't know about that macro.

Reply:
For me, this macro is pure black magic... but oops, I picked the wrong macro, I was thinking at: ``_Py_SIZE_ROUND_UP()``.

```
/* Below "a" is a power of 2. */
/* Round down size "n" to be a multiple of "a". */
#define _Py_SIZE_ROUND_DOWN(n, a) ((size_t)(n) & ~(size_t)((a) - 1))

/* Round up size "n" to be a multiple of "a". */
#define _Py_SIZE_ROUND_UP(n, a) (((size_t)(n) + \
        (size_t)((a) - 1)) & ~(size_t)((a) - 1))
```

Reply:
Maybe _Py_SIZE_ROUND_UP() should be rewritten as a static inline function to ensure that (a-1) expression is only computed once.

Reply:
We do not want to round to a multiple of anything (even a power of two), we want to round up to a pure power of two.